lib/internal/errors.js

```diff
@@ -0,0 +1,98 @@
+'use strict';
+
+// The whole point behind this internal module is to allow Node.js to no
+// longer be forced to treat every error message change as a semver-major
+// change. The NodeError classes here all expose a `code` property whose
+// value statically and permanently identifies the error. While the error
+// message may change, the code should not.
+
+const kCode = Symbol('code');
+const kSymbolsMap = new Map();
+
+function symbolFor(name, add) {
+  var sym = kSymbolsMap.get(name);
+  if (add && sym === undefined) {
+    sym = Symbol(name);
+    kSymbolsMap.set(name, sym);
+  }
+  return sym;
+}
+
+var assert, util;
+function lazyAssert() {
+  if (!assert)
+    assert = require('assert');
+  return assert;
+}
+
+function lazyUtil() {
+  if (!util)
+    util = require('util');
+  return util;
+}
+
+function makeNodeError(Base) {
+  return class NodeError extends Base {
+    constructor(key, ...args) {
+      super(message(key, args));
+      this[kCode] = key;
+      Error.captureStackTrace(this, NodeError);
+    }
+
+    get name() {
+      return `${super.name}[${this[kCode]}]`;
+    }
+
+    get code() {
+      return this[kCode];
+    }
+  };
+}
+
+function message(key, args) {
+  const assert = lazyAssert();
+  assert.strictEqual(typeof key, 'string');
+  const util = lazyUtil();
+  const msg = exports[symbolFor(key)];
+  assert(msg, `An invalid error message key was used: ${key}.`);
+  let fmt = util.format;
+  if (typeof msg === 'function') {
+    fmt = msg;
+  } else {
+    if (args === undefined || args.length === 0)
+      return msg;
+    args.unshift(msg);
+  }
+  return String(fmt.apply(null, args));
```

Comment:
`fmt(...args);`? Not sure whether that was part of your conversation with @joyeecheung or not